pkg/provider/provider.go

```diff
@@ -525,218 +593,136 @@ func DSN(
+			return nil, fmt.Errorf("invalid authenticator %s", authenticator)
 		}
 	}
 
-	if privateKeyPath != "" { //nolint:gocritic // todo: please fix this to pass gocritic
-		privateKeyBytes, err := ReadPrivateKeyFile(privateKeyPath)
-		if err != nil {
-			return "", fmt.Errorf("private Key file could not be read err = %w", err)
-		}
-		rsaPrivateKey, err := ParsePrivateKey(privateKeyBytes, []byte(privateKeyPassphrase))
-		if err != nil {
-			return "", fmt.Errorf("private Key could not be parsed err = %w", err)
-		}
-		config.PrivateKey = rsaPrivateKey
-		config.Authenticator = gosnowflake.AuthTypeJwt
-	} else if privateKey != "" {
-		rsaPrivateKey, err := ParsePrivateKey([]byte(privateKey), []byte(privateKeyPassphrase))
+	if v, ok := s.GetOk("passcode"); ok && v.(string) != "" {
+		config.Passcode = v.(string)
+	}
+
+	if v, ok := s.GetOk("passcode_in_password"); ok && v.(bool) {
+		config.PasscodeInPassword = v.(bool)
+	}
+	if v, ok := s.GetOk("okta_url"); ok && v.(string) != "" {
+		oktaURL, err := url.Parse(v.(string))
 		if err != nil {
-			return "", fmt.Errorf("private Key could not be parsed err = %w", err)
-		}
-		config.PrivateKey = rsaPrivateKey
-		config.Authenticator = gosnowflake.AuthTypeJwt
-	} else if browserAuth {
-		config.Authenticator = gosnowflake.AuthTypeExternalBrowser
-	} else if oauthAccessToken != "" {
-		config.Authenticator = gosnowflake.AuthTypeOAuth
-		config.Token = oauthAccessToken
-	} else if password != "" {
-		config.Password = password
-	} else if account == "" || user == "" {
-		// If account and user are empty then we need to fall back on using profile config
-		log.Printf("[DEBUG] No account or user provided, falling back to profile %s\n", profile)
-		if profile == "default" {
-			defaultConfig := sdk.DefaultConfig()
-			if defaultConfig.Account == "" || defaultConfig.User == "" {
-				return "", errors.New("account and User must be set in provider config, ~/.snowflake/config, or as an environment variable")
-			}
-			config = sdk.MergeConfig(config, defaultConfig)
-		} else {
-			profileConfig, err := sdk.ProfileConfig(profile)
-			if err != nil {
-				return "", errors.New("could not retrieve profile config: " + err.Error())
-			}
-			if profileConfig == nil {
-				return "", errors.New("profile with name: " + profile + " not found in config file")
-			}
-			// merge any credentials found in profile with config
-			config = sdk.MergeConfig(config, profileConfig)
+			return nil, fmt.Errorf("could not parse okta_url err = %w", err)
 		}
+		config.OktaURL = oktaURL
 	}
-	config.Application = "terraform-provider-snowflake"
 
-	return gosnowflake.DSN(config)
-}
-
-func ReadPrivateKeyFile(privateKeyPath string) ([]byte, error) {
-	expandedPrivateKeyPath, err := homedir.Expand(privateKeyPath)
-	if err != nil {
-		return nil, fmt.Errorf("invalid Path to private key err = %w", err)
+	if v, ok := s.GetOk("login_timeout"); ok {
+		config.LoginTimeout = time.Duration(int64(v.(int)))
```

Comment:
I think every conversion to time.Duration is incorrect. time.Duration as a single value represents nano second. In the documentation you've mentioned you expect users to pass seconds. Then it should be assigned like this
```go
config.LoginTimeout = time.Second * time.Duration(v.(int))
```

Reply:
fixed